### PR TITLE
add AwaitableStream prototype

### PR DESCRIPTION
**Pull request by anurse**
This is just moving @davidfowl 's prototype code over here. I'd like to merge this fairly soon, but we don't know when it will ship, so I'm leaving it at 0.1 for now.

/cc @davidfowl @halter73 @Eilon @muratg

### CODE REVIEW

**Comment by davidfowl** [comment]
:shipit: !

File: src/Microsoft.Extensions.AwaitableStream/AwaitableStream.cs [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+                _tail = segment;
+            }
+
+            // Call the continuation
+            Complete();
+
+            if (!_consumeCalled)
+            {
+                // Call it on the user's behalf
+                Consumed(count);
```
Comment (halter73) [inline src/Microsoft.Extensions.AwaitableStream/AwaitableStream.cs]: Can we make it do the auto-consume on next `ReadAsync` before merging?

**Comment by halter73** [comment]
What do we think about distinguishing consumed and examined for this?

**Comment by davidfowl** [comment]
No. Lets merge the initial changes and tests and iterate.

File: test/Microsoft.Extensions.AwaitableStream.Test/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+// associated with an assembly.
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("Microsoft.Extensions.AwaitableStream.Test")]
+[assembly: AssemblyTrademark("")]
+
+// Setting ComVisible to false makes the types in this assembly not visible
+// to COM components.  If you need to access a type in this assembly from
+// COM, set the ComVisible attribute to true on that type.
+[assembly: ComVisible(false)]
```
Comment (natemcmaster) [inline test/Microsoft.Extensions.AwaitableStream.Test/Properties/AssemblyInfo.cs]: Can be removed

Comment (davidfowl) [inline test/Microsoft.Extensions.AwaitableStream.Test/Properties/AssemblyInfo.cs]: But what if we want to make it comvisible in the future 😄

**Comment by anurse** [comment]
Yeah, I'm with Fowler. Iterating is good. I'm already working on the Read/Write synchronization (and I can work in the "consume on ReadAsync" stuff too).

**Pull request merged**